Workspace symbol picker doesn't seem to work with Intelephense

### Summary

Using [Intelephense](https://intelephense.com/), workspace symbol lookup doesn't seem to work. Works fine in VS Code.

### Reproduction Steps

I tried this:

Made a new directory, ran `composer init`, chose the namespace `TestHelix`, made two files in the `src` directory like so:
```
.
├── composer.json
├── src
│   ├── Bar.php
│   └── Foo.php
└── vendor
    ├── autoload.php
    └── composer
        ├── autoload_classmap.php
        ├── autoload_namespaces.php
        ├── autoload_psr4.php
        ├── autoload_real.php
        ├── autoload_static.php
        ├── ClassLoader.php
        └── LICENSE
```
With the following contents:
<details><summary>composer.json</summary>

```json
{
    "name": "ejulen/test-hx-php",
    "autoload": {
        "psr-4": {
            "TestHelix\\": "src/"
        }
    },
    "require": {}
}
```
</details>
<details><summary>Foo.php</summary>

```php
<?php

namespace TestHelix;

class Foo {}
```
</details>
<details><summary>Bar.php</summary>

```php
<?php

namespace TestHelix;

class Bar {}
```
</details>

Then:
1. `hx -vvv src/Foo.php`
2. `<space>S`

I expected this to happen:
For a list of workspace symbols to show up.

Instead, this happened:
An empty picker shows up.


### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-08-31T20:34:19.777 helix_loader [DEBUG] Located configuration folders: []
2022-08-31T20:34:19.781 helix_view::clipboard::provider [WARN] No clipboard provider found! Yanking and pasting will be internal to Helix
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE | WRITABLE
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(2), interests=READABLE | WRITABLE
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(3), interests=READABLE | WRITABLE
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(4), interests=READABLE | WRITABLE
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(5), interests=READABLE | WRITABLE
2022-08-31T20:34:19.801 helix_lsp::client [INFO] Using custom LSP config: {}
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(0), interests=READABLE
2022-08-31T20:34:19.801 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialize","params":{"capabilities":{"textDocument":{"codeAction":{"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}}},"completion":{"completionItem":{"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"snippetSupport":false},"completionItemKind":{}},"hover":{"contentFormat":["markdown"]},"publishDiagnostics":{},"rename":{"dynamicRegistration":false,"honorsChangeAnnotations":false,"prepareSupport":false},"signatureHelp":{"signatureInformation":{"activeParameterSupport":true,"documentationFormat":["markdown"],"parameterInformation":{"labelOffsetSupport":true}}}},"window":{"workDoneProgress":true},"workspace":{"applyEdit":true,"configuration":true,"didChangeConfiguration":{"dynamicRegistration":false},"symbol":{"dynamicRegistration":false},"workspaceFolders":true}},"initializationOptions":{},"processId":1135379,"rootPath":"/home/johannes/dev/test-hx-php","rootUri":"file:///home/johannes/dev/test-hx-php","workspaceFolders":[{"name":"test-hx-php","uri":"file:///home/johannes/dev/test-hx-php"}]},"id":0}
2022-08-31T20:34:19.801 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE
2022-08-31T20:34:20.133 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Initialising intelephense 1.8.2"}}
2022-08-31T20:34:20.133 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Initialising intelephense 1.8.2" }
2022-08-31T20:34:20.135 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Reading state from /tmp/intelephense/4e5e2af1."}}
2022-08-31T20:34:20.135 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Reading state from /tmp/intelephense/4e5e2af1." }
2022-08-31T20:34:20.269 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Intelephense premium features enabled."}}
2022-08-31T20:34:20.269 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Initialised in 137 ms"}}
2022-08-31T20:34:20.269 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Intelephense premium features enabled." }
2022-08-31T20:34:20.269 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"textDocumentSync":2,"documentSymbolProvider":true,"workspaceSymbolProvider":true,"completionProvider":{"triggerCharacters":["$",">",":","\\","/","*",".","<"],"resolveProvider":true},"signatureHelpProvider":{"triggerCharacters":["(",",",":"]},"definitionProvider":true,"documentFormattingProvider":true,"documentRangeFormattingProvider":true,"referencesProvider":true,"hoverProvider":true,"documentHighlightProvider":true,"foldingRangeProvider":true,"implementationProvider":true,"declarationProvider":true,"workspace":{"workspaceFolders":{"supported":true,"changeNotifications":true}},"renameProvider":{"prepareProvider":true},"typeDefinitionProvider":true,"selectionRangeProvider":true,"codeActionProvider":true,"executeCommandProvider":{"commands":["intelephense.import.symbol","intelephense.implement.abstract.method.all","intelephense.phpdoc.add"]}}}}
2022-08-31T20:34:20.269 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Initialised in 137 ms" }
2022-08-31T20:34:20.269 helix_lsp::transport [INFO] <- {"capabilities":{"codeActionProvider":true,"completionProvider":{"resolveProvider":true,"triggerCharacters":["$",">",":","\\","/","*",".","<"]},"declarationProvider":true,"definitionProvider":true,"documentFormattingProvider":true,"documentHighlightProvider":true,"documentRangeFormattingProvider":true,"documentSymbolProvider":true,"executeCommandProvider":{"commands":["intelephense.import.symbol","intelephense.implement.abstract.method.all","intelephense.phpdoc.add"]},"foldingRangeProvider":true,"hoverProvider":true,"implementationProvider":true,"referencesProvider":true,"renameProvider":{"prepareProvider":true},"selectionRangeProvider":true,"signatureHelpProvider":{"triggerCharacters":["(",",",":"]},"textDocumentSync":2,"typeDefinitionProvider":true,"workspace":{"workspaceFolders":{"changeNotifications":true,"supported":true}},"workspaceSymbolProvider":true}}
2022-08-31T20:34:20.269 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"initialized","params":{}}
2022-08-31T20:34:20.270 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"workspace/didChangeConfiguration","params":{"settings":{}}}
2022-08-31T20:34:20.270 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"languageId":"php","text":"<?php\n\nnamespace TestHelix;\n\nclass Foo {}\n","uri":"file:///home/johannes/dev/test-hx-php/src/Foo.php","version":0}}}
2022-08-31T20:34:20.271 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":0,"method":"workspace/configuration","params":{"items":[{"section":"intelephense"},{"section":"intelephense","scopeUri":"file:///home/johannes/dev/test-hx-php"}]}}
2022-08-31T20:34:20.271 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","result":[null,null],"id":0}
2022-08-31T20:34:20.291 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Searching file:///home/johannes/.fnm/node-versions/v16.15.0/installation/lib/node_modules/intelephense/lib/stub for files to index."}}
2022-08-31T20:34:20.291 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Searching file:///home/johannes/.fnm/node-versions/v16.15.0/installation/lib/node_modules/intelephense/lib/stub for files to index." }
2022-08-31T20:34:20.304 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Searching file:///home/johannes/dev/test-hx-php for files to index."}}
2022-08-31T20:34:20.304 helix_term::application [INFO] window/logMessage: LogMessageParams { typ: Info, message: "Searching file:///home/johannes/dev/test-hx-php for files to index." }
2022-08-31T20:34:20.783 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/johannes/dev/test-hx-php/src/Foo.php","diagnostics":[]}}
2022-08-31T20:34:22.311 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///home/johannes/dev/test-hx-php/src/Foo.php","diagnostics":[]}}
2022-08-31T20:34:25.515 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"workspace/symbol","params":{"query":""},"id":1}
2022-08-31T20:34:25.516 helix_lsp::transport [INFO] <- {"jsonrpc":"2.0","id":1,"result":[]}
2022-08-31T20:34:25.516 helix_lsp::transport [INFO] <- []
2022-08-31T20:34:27.385 helix_term::ui::picker [DEBUG] picker score 1.342µs
2022-08-31T20:34:27.644 helix_term::ui::picker [DEBUG] picker score 3.346µs
2022-08-31T20:34:27.713 helix_term::ui::picker [DEBUG] picker score 1.372µs
2022-08-31T20:34:30.951 helix_term::job [DEBUG] waiting on jobs...
2022-08-31T20:34:30.951 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"shutdown","params":null,"id":2}
2022-08-31T20:34:31.453 helix_term::application [ERROR] Timed out waiting for language servers to shutdown
2022-08-31T20:34:31.453 mio::poll [TRACE] deregistering event source from poller
2022-08-31T20:34:31.454 mio::poll [TRACE] deregistering event source from poller
2022-08-31T20:34:31.454 mio::poll [TRACE] deregistering event source from poller
2022-08-31T20:34:31.454 mio::poll [TRACE] deregistering event source from poller
2022-08-31T20:34:31.454 mio::poll [TRACE] deregistering event source from poller
2022-08-31T20:34:31.454 mio::poll [TRACE] deregistering event source from poller
```

</details>


### Platform

Linux

### Terminal Emulator

Windows Terminal (SSH)

### Helix Version

helix 22.08

--- a/helix-term/src/commands/lsp.rs
+++ b/helix-term/src/commands/lsp.rs
@@ -1,3 +1,4 @@
+use futures_util::FutureExt;
 use helix_lsp::{
     block_on,
     lsp::{self, CodeAction, CodeActionOrCommand, DiagnosticSeverity, NumberOrString},
@@ -14,7 +15,8 @@ use helix_view::{apply_transaction, document::Mode, editor::Action, theme::Style
 use crate::{
     compositor::{self, Compositor},
     ui::{
-        self, lsp::SignatureHelp, overlay::overlayed, FileLocation, FilePicker, Popup, PromptEvent,
+        self, lsp::SignatureHelp, overlay::overlayed, DynamicPicker, FileLocation, FilePicker,
+        Popup, PromptEvent,
     },
 };
 
@@ -384,10 +386,48 @@ pub fn workspace_symbol_picker(cx: &mut Context) {
     cx.callback(
         future,
         move |_editor, compositor, response: Option<Vec<lsp::SymbolInformation>>| {
-            if let Some(symbols) = response {
-                let picker = sym_picker(symbols, current_url, offset_encoding);
-                compositor.push(Box::new(overlayed(picker)))
-            }
+            let symbols = match response {
+                Some(s) => s,
+                None => return,
+            };
+            let picker = sym_picker(symbols, current_url, offset_encoding);
+            let get_symbols = |query: String, editor: &mut Editor| {
+                let doc = doc!(editor);
+                let language_server = match doc.language_server() {
+                    Some(s) => s,
+                    None => {
+                        // This should not generally happen since the picker will not
+                        // even open in the first place if there is no server.
+                        return async move { Err(anyhow::anyhow!("LSP not active")) }.boxed();
+                    }
+                };
+                let symbol_request = match language_server.workspace_symbols(query) {
+                    Some(future) => future,
+                    None => {
+                        // This should also not happen since the language server must have
+                        // supported workspace symbols before to reach this block.
+                        return async move {
+                            Err(anyhow::anyhow!(
+                                "Language server does not support workspace symbols"
+                            ))
+                        }
+                        .boxed();
+                    }
+                };
+
+                let future = async move {
+                    let json = symbol_request.await?;
+                    let response: Option<Vec<lsp::SymbolInformation>> =
+                        serde_json::from_value(json)?;
+
+                    response.ok_or_else(|| {
+                        anyhow::anyhow!("No response for workspace symbols from language server")
+                    })
+                };
+                future.boxed()
+            };
+            let dyn_picker = DynamicPicker::new(picker, Box::new(get_symbols));
+            compositor.push(Box::new(overlayed(dyn_picker)))
         },
     )
 }

--- a/helix-term/src/commands/lsp.rs
+++ b/helix-term/src/commands/lsp.rs
@@ -386,10 +386,7 @@ pub fn workspace_symbol_picker(cx: &mut Context) {
     cx.callback(
         future,
         move |_editor, compositor, response: Option<Vec<lsp::SymbolInformation>>| {
-            let symbols = match response {
-                Some(s) => s,
-                None => return,
-            };
+            let symbols = response.unwrap_or_default();
             let picker = sym_picker(symbols, current_url, offset_encoding);
             let get_symbols = |query: String, editor: &mut Editor| {
                 let doc = doc!(editor);
@@ -420,9 +417,7 @@ pub fn workspace_symbol_picker(cx: &mut Context) {
                     let response: Option<Vec<lsp::SymbolInformation>> =
                         serde_json::from_value(json)?;
 
-                    response.ok_or_else(|| {
-                        anyhow::anyhow!("No response for workspace symbols from language server")
-                    })
+                    Ok(response.unwrap_or_default())
                 };
                 future.boxed()
             };

--- a/helix-term/src/ui/mod.rs
+++ b/helix-term/src/ui/mod.rs
@@ -19,7 +19,7 @@ pub use completion::Completion;
 pub use editor::EditorView;
 pub use markdown::Markdown;
 pub use menu::Menu;
-pub use picker::{FileLocation, FilePicker, Picker};
+pub use picker::{DynamicPicker, FileLocation, FilePicker, Picker};
 pub use popup::Popup;
 pub use prompt::{Prompt, PromptEvent};
 pub use spinner::{ProgressSpinners, Spinner};

--- a/helix-term/src/ui/overlay.rs
+++ b/helix-term/src/ui/overlay.rs
@@ -69,4 +69,8 @@ impl<T: Component + 'static> Component for Overlay<T> {
         let dimensions = (self.calc_child_size)(area);
         self.content.cursor(dimensions, ctx)
     }
+
+    fn id(&self) -> Option<&'static str> {
+        self.content.id()
+    }
 }

--- a/helix-term/src/ui/picker.rs
+++ b/helix-term/src/ui/picker.rs
@@ -469,32 +469,39 @@ impl<T: Item> Picker<T> {
 
             self.matches.sort_unstable();
         } else {
-            let query = FuzzyQuery::new(pattern);
-            self.matches.clear();
-            self.matches.extend(
-                self.options
-                    .iter()
-                    .enumerate()
-                    .filter_map(|(index, option)| {
-                        let text = option.filter_text(&self.editor_data);
-
-                        query
-                            .fuzzy_match(&text, &self.matcher)
-                            .map(|score| PickerMatch {
-                                index,
-                                score,
-                                len: text.chars().count(),
-                            })
-                    }),
-            );
-            self.matches.sort_unstable();
+            self.force_score();
         }
 
         log::debug!("picker score {:?}", Instant::now().duration_since(now));
 
         // reset cursor position
         self.cursor = 0;
+        let pattern = self.prompt.line();
         self.previous_pattern.clone_from(pattern);
+    }
+
+    pub fn force_score(&mut self) {
+        let pattern = self.prompt.line();
+
+        let query = FuzzyQuery::new(pattern);
+        self.matches.clear();
+        self.matches.extend(
+            self.options
+                .iter()
+                .enumerate()
+                .filter_map(|(index, option)| {
+                    let text = option.filter_text(&self.editor_data);
+
+                    query
+                        .fuzzy_match(&text, &self.matcher)
+                        .map(|score| PickerMatch {
+                            index,
+                            score,
+                            len: text.chars().count(),
+                        })
+                }),
+        );
+        self.matches.sort_unstable();
     }
 
     /// Move the cursor by a number of lines, either down (`Forward`) or up (`Backward`)

--- a/helix-term/src/ui/picker.rs
+++ b/helix-term/src/ui/picker.rs
@@ -3,6 +3,7 @@ use crate::{
     ctrl, key, shift,
     ui::{self, fuzzy_match::FuzzyQuery, EditorView},
 };
+use futures_util::future::BoxFuture;
 use tui::{
     buffer::Buffer as Surface,
     widgets::{Block, BorderType, Borders},
@@ -22,7 +23,7 @@ use helix_view::{
     Document, DocumentId, Editor,
 };
 
-use super::menu::Item;
+use super::{menu::Item, overlay::Overlay};
 
 pub const MIN_AREA_WIDTH_FOR_PREVIEW: u16 = 72;
 /// Biggest file size to preview in bytes
@@ -749,5 +750,76 @@ impl<T: Item + 'static> Component for Picker<T> {
         let area = inner.clip_left(1).with_height(1);
 
         self.prompt.cursor(area, editor)
+    }
+}
+
+/// Returns a new list of options to replace the contents of the picker
+/// when called with the current picker query,
+pub type DynQueryCallback<T> =
+    Box<dyn Fn(String, &mut Editor) -> BoxFuture<'static, anyhow::Result<Vec<T>>>>;
+
+/// A picker that updates its contents via a callback whenever the
+/// query string changes. Useful for live grep, workspace symbols, etc.
+pub struct DynamicPicker<T: ui::menu::Item + Send> {
+    file_picker: FilePicker<T>,
+    query_callback: DynQueryCallback<T>,
+}
+
+impl<T: ui::menu::Item + Send> DynamicPicker<T> {
+    pub const ID: &'static str = "dynamic-picker";
+
+    pub fn new(file_picker: FilePicker<T>, query_callback: DynQueryCallback<T>) -> Self {
+        Self {
+            file_picker,
+            query_callback,
+        }
+    }
+}
+
+impl<T: Item + Send + 'static> Component for DynamicPicker<T> {
+    fn render(&mut self, area: Rect, surface: &mut Surface, cx: &mut Context) {
+        self.file_picker.render(area, surface, cx);
+    }
+
+    fn handle_event(&mut self, event: &Event, cx: &mut Context) -> EventResult {
+        let prev_query = self.file_picker.picker.prompt.line().to_owned();
+        let event_result = self.file_picker.handle_event(event, cx);
+        let current_query = self.file_picker.picker.prompt.line();
+
+        if *current_query == prev_query || matches!(event_result, EventResult::Ignored(_)) {
+            return event_result;
+        }
+
+        let new_options = (self.query_callback)(current_query.to_owned(), cx.editor);
+
+        cx.jobs.callback(async move {
+            let new_options = new_options.await?;
+            let callback =
+                crate::job::Callback::EditorCompositor(Box::new(move |_editor, compositor| {
+                    // Wrapping of pickers in overlay is done outside the picker code,
+                    // so this is fragile and will break if wrapped in some other widget.
+                    let picker = match compositor.find_id::<Overlay<DynamicPicker<T>>>(Self::ID) {
+                        Some(overlay) => &mut overlay.content.file_picker.picker,
+                        None => return,
+                    };
+                    picker.options = new_options;
+                    picker.cursor = 0;
+                    picker.force_score();
+                }));
+            anyhow::Ok(callback)
+        });
+        EventResult::Consumed(None)
+    }
+
+    fn cursor(&self, area: Rect, ctx: &Editor) -> (Option<Position>, CursorKind) {
+        self.file_picker.cursor(area, ctx)
+    }
+
+    fn required_size(&mut self, viewport: (u16, u16)) -> Option<(u16, u16)> {
+        self.file_picker.required_size(viewport)
+    }
+
+    fn id(&self) -> Option<&'static str> {
+        Some(Self::ID)
     }
 }

--- a/helix-term/src/ui/picker.rs
+++ b/helix-term/src/ui/picker.rs
@@ -798,7 +798,7 @@ impl<T: Item + Send + 'static> Component for DynamicPicker<T> {
         cx.jobs.callback(async move {
             let new_options = new_options.await?;
             let callback =
-                crate::job::Callback::EditorCompositor(Box::new(move |_editor, compositor| {
+                crate::job::Callback::EditorCompositor(Box::new(move |editor, compositor| {
                     // Wrapping of pickers in overlay is done outside the picker code,
                     // so this is fragile and will break if wrapped in some other widget.
                     let picker = match compositor.find_id::<Overlay<DynamicPicker<T>>>(Self::ID) {
@@ -808,6 +808,7 @@ impl<T: Item + Send + 'static> Component for DynamicPicker<T> {
                     picker.options = new_options;
                     picker.cursor = 0;
                     picker.force_score();
+                    editor.reset_idle_timer();
                 }));
             anyhow::Ok(callback)
         });

--- a/helix-term/src/ui/picker.rs
+++ b/helix-term/src/ui/picker.rs
@@ -763,6 +763,7 @@ pub type DynQueryCallback<T> =
 pub struct DynamicPicker<T: ui::menu::Item + Send> {
     file_picker: FilePicker<T>,
     query_callback: DynQueryCallback<T>,
+    query: String,
 }
 
 impl<T: ui::menu::Item + Send> DynamicPicker<T> {
@@ -772,6 +773,7 @@ impl<T: ui::menu::Item + Send> DynamicPicker<T> {
         Self {
             file_picker,
             query_callback,
+            query: String::new(),
         }
     }
 }
@@ -782,13 +784,14 @@ impl<T: Item + Send + 'static> Component for DynamicPicker<T> {
     }
 
     fn handle_event(&mut self, event: &Event, cx: &mut Context) -> EventResult {
-        let prev_query = self.file_picker.picker.prompt.line().to_owned();
         let event_result = self.file_picker.handle_event(event, cx);
         let current_query = self.file_picker.picker.prompt.line();
 
-        if *current_query == prev_query || matches!(event_result, EventResult::Ignored(_)) {
+        if !matches!(event, Event::IdleTimeout) || self.query == *current_query {
             return event_result;
         }
+
+        self.query.clone_from(current_query);
 
         let new_options = (self.query_callback)(current_query.to_owned(), cx.editor);
 

